R/lint.R

```diff
@@ -82,67 +82,107 @@ get_usage_function_names <- function(x) {
+
+  check_installed(c("lintr", "vctrs", "xml2"))
+
+  lintr::lint_dir(
+    path = path,
+    pattern = pattern,
+    linters = lifecycle_linter(packages = packages, which = which, symbol_is_undesirable = symbol_is_undesirable)
+  )
+}
+
+#' @rdname lifecycle_linter
+#' @export
+lint_tidyverse_lifecycle <- function(
+    path = ".",
+    pattern = "(?i)[.](r|rmd|qmd|rnw|rhtml|rrst|rtex|rtxt)$",
+    which = c("superseded", "deprecated", "questioning", "defunct", "experimental", "soft-deprecated", "retired"),
+    symbol_is_undesirable = FALSE
+) {
+  which <- match.arg(which, several.ok = TRUE)
+
+  check_installed(c("lintr", "vctrs", "xml2", "tidyverse"))
+
+  lint_lifecycle(
+    packages = tidyverse::tidyverse_packages(),
+    pattern = pattern,
+    path = path,
+    which = which,
+    symbol_is_undesirable = symbol_is_undesirable
+  )
+}
+
 #' Lint usages of functions that have a non-stable life cycle.
 #'
-#' - `lint_lifecycle` dynamically queries the package documentation for packages
+#' - `lifecycle_linter()` creates a linter for lifecycle annotations which can be
+#'   included in a `.lintr` configuration if `lintr` is used directly.
+#' - `lint_lifecycle()` dynamically queries the package documentation for packages
 #'   in `packages` for lifecycle annotations and then searches the directory in
 #'   `path` for usages of those functions.
-#' - `lint_tidyverse_lifecycle` is a convenience function to call `lint_lifecycle`
+#' - `lint_tidyverse_lifecycle()` is a convenience function to call `lint_lifecycle()`
 #'   for all the packages in the tidyverse.
-#' - `pkg_lifecycle_statuses` returns a data frame of functions with lifecycle
+#' - `pkg_lifecycle_statuses()` returns a data frame of functions with lifecycle
 #'   annotations for an installed package.
 #'
 #' @param packages One or more installed packages to query for lifecycle statuses.
-#' @param path The directory path to the files you want to search.
-#' @param pattern Any files matching this pattern will be searched. The default
-#'   searches any files ending in `.R` or `.Rmd`.
+#' @param which Vector of lifecycle statuses to lint.
+#' @param symbol_is_undesirable Also lint symbol usages, e.g. `lapply(x, is_na)`?
+#'
+#' @examples
+#'
+#' lintr::lint(text = "is_na(x)", linters = lifecycle_linter(packages = "rlang"))
+#' lintr::lint(text = "lapply(x, is_na)", linters = lifecycle_linter(packages = "rlang", symbol_is_undesirable = TRUE))
+#'
 #' @export
-lint_lifecycle <- function(packages, path = ".", pattern = "[.][Rr](md)?", which = c("superseded", "deprecated", "questioning", "defunct", "experimental", "soft-deprecated", "retired")) {
-  which <- match.arg(which, several.ok = TRUE)
-
-  check_installed(c("lintr", "vctrs"))
+lifecycle_linter <- function(
+    packages = tidyverse::tidyverse_packages(),
+    which = c("superseded", "deprecated", "questioning", "defunct", "experimental", "soft-deprecated", "retired"),
+    symbol_is_undesirable = FALSE
+) {
```

Comment:
Idea for future expansion: `exclude = character()` arg to explicitly allow some functions.